nvflare/private/fed/server/job_meta_validator.py

```diff
@@ -0,0 +1,201 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import collections
+import json
+import logging
+from io import BytesIO
+from typing import Optional, Set, Tuple
+from zipfile import ZipFile
+
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.job_def import JobMetaKey
+
+SERVER_CONFIG = "config_fed_server.json"
+CLIENT_CONFIG = "config_fed_client.json"
+META = "meta.json"
+ALL_SITES = "@ALL"
+MAX_CLIENTS = 1000000
+
+logger = logging.getLogger(__name__)
+
+
+class JobMetaValidator:
+    """Job validator"""
+
+    def __init__(self, fl_ctx: FLContext):
+        self.fl_ctx = fl_ctx
+
+    def validate(self, job_name: str, job_data: bytes) -> Tuple[bool, str, dict]:
+        """Validate job
+
+        Args:
+            job_name (str): Job name
+            job_data (bytes): Job ZIP data
+
+        Returns:
+            Tuple[bool, str, dict]: (is_valid, error_message, meta)
+        """
+
+        meta = {}
+        try:
+            with ZipFile(BytesIO(job_data), "r") as zf:
+                meta = self._validate_meta(job_name, zf)
+                site_list = self._validate_deploy_map(job_name, meta)
+                self._validate_app(job_name, meta, zf)
+                clients = self._get_all_clients(site_list)
+                self._validate_min_clients(job_name, meta, clients)
+                self._validate_resource(job_name, meta)
+                self._validate_mandatory_clients(job_name, meta, clients)
+
+        except ValueError as e:
+            return False, str(e), meta
+
+        return True, "", meta
+
+    @staticmethod
+    def _validate_meta(job_name: str, zf: ZipFile) -> Optional[dict]:
+        meta_file = f"{job_name}/{META}"
+        logger.debug(f"validate file {meta_file} exists for job {job_name}")
+        meta = None
+
+        if meta_file in zf.namelist():
+            meta_data = zf.read(meta_file)
+            meta = json.loads(meta_data)
+        return meta
+
+    @staticmethod
+    def _validate_deploy_map(job_name: str, meta: dict) -> list:
+
+        if not meta:
+            raise ValueError(f"meta.json is empty for job {job_name}")
+
+        deploy_map = meta.get(JobMetaKey.DEPLOY_MAP.value)
+        if not deploy_map:
+            raise ValueError(f"deploy_map is empty for job {job_name}")
+
+        site_list = []
+        for app, deployment in deploy_map.items():
+            if deployment:
+                site_list.extend(deployment)
+
+        if not site_list:
+            raise ValueError(f"No site is specified in deploy_map for job {job_name}")
+
+        if ALL_SITES.casefold() in (site.casefold() for site in site_list):
+            # if @ALL is specified, no other site can be in the list
+            if len(site_list) > 1:
+                raise ValueError(f"No other site can be specified if @ALL is used for job {job_name}")
+            else:
+                site_list = [ALL_SITES]
```

Comment:
else is unnecessary, as ALL_SITES is already in the site_list

Reply:
The purpose of this line is to use canonical form (all caps) in site_list so comparison in other code is easier.